flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -0,0 +1,347 @@
+-->
+
+void TableColumn(TableCreationContext context) :
+{
+}
+{
+    (
+        TableColumn2(context.columnList)
+    |
+        context.watermark = Watermark()
+    |
+        context.primaryKeyList = PrimaryKey()
+    |
+        UniqueKey(context.uniqueKeysList)
+    |
+        ComputedColumn(context)
+    )
+}
+
+void ComputedColumn(TableCreationContext context) :
+{
+    SqlNode identifier;
+    SqlNode expr;
+    boolean hidden = false;
+    SqlParserPos pos;
+}
+{
+    identifier = SimpleIdentifier() {pos = getPos();}
+    <AS>
+    expr = Expression(ExprContext.ACCEPT_SUB_QUERY) {
+        expr = SqlStdOperatorTable.AS.createCall(Span.of(identifier, expr).pos(), expr, identifier);
+        context.columnList.add(expr);
+    }
+}
+
+void TableColumn2(List<SqlNode> list) :
+{
+    SqlParserPos pos;
+    SqlIdentifier name;
+    SqlDataTypeSpec type;
+    SqlCharStringLiteral comment = null;
+    boolean isHeader = false;
+}
+{
+    name = SimpleIdentifier()
+    type = DataType()
+    (
+        <NULL> { type = type.withNullable(true); }
+    |
+        <NOT> <NULL> { type = type.withNullable(false); }
+    |
+        { type = type.withNullable(true); }
+    )
+    [ <HEADER> { isHeader = true; } ]
+    [ <COMMENT> <QUOTED_STRING> {
+        String p = SqlParserUtil.parseString(token.image);
+        comment = SqlLiteral.createCharString(p, getPos());
+    }]
+    {
+        SqlTableColumn tableColumn = new SqlTableColumn(name, type, comment, getPos());
+        tableColumn.setHeader(isHeader);
+        list.add(tableColumn);
+    }
+}
+
+SqlNodeList PrimaryKey() :
+{
+    List<SqlNode> pkList = new ArrayList<SqlNode>();
+
+    SqlParserPos pos;
+    SqlIdentifier columnName;
+}
+{
+    <PRIMARY> { pos = getPos(); } <KEY> <LPAREN>
+        columnName = SimpleIdentifier() { pkList.add(columnName); }
+        (<COMMA> columnName = SimpleIdentifier() { pkList.add(columnName); })*
+    <RPAREN>
+    {
+        return new SqlNodeList(pkList, pos.plus(getPos()));
+    }
+}
+
+void UniqueKey(List<SqlNodeList> list) :
+{
+    List<SqlNode> ukList = new ArrayList<SqlNode>();
+    SqlParserPos pos;
+    SqlIdentifier columnName;
+}
+{
+    <UNIQUE> { pos = getPos(); } <LPAREN>
+        columnName = SimpleIdentifier() { ukList.add(columnName); }
+        (<COMMA> columnName = SimpleIdentifier() { ukList.add(columnName); })*
+    <RPAREN>
+    {
+        SqlNodeList uk = new SqlNodeList(ukList, pos.plus(getPos()));
+        list.add(uk);
+    }
+}
+
+SqlWatermark Watermark() :
+{
+    SqlIdentifier watermarkName = null;
+    SqlIdentifier columnName;
+    SqlWatermarkStrategy strategy;
+    int delayDef = -1;
+    SqlTimeUnit timeUnit = null;
+}
+{
+    <WATERMARK>
+    [watermarkName = SimpleIdentifier()]
+    <FOR>
+    columnName = SimpleIdentifier()
+    <AS>
+    (
+        <BOUNDED> <WITH> <DELAY>
+            {
+                strategy = SqlWatermarkStrategy.BOUNDED_WITH_DELAY;
+                delayDef = UnsignedIntLiteral();
```

Comment:
The interval literal should be surrounded in quotes according to the design doc?

Reply:
There is no necessity to quoting the literal.

Reply:
We want to keep closer to standard for the new introduced syntax. The interval time unit syntax quotes the literal. See https://calcite.apache.org/docs/reference.html